### ISSUE TITLE
Calling StateManager.SetStateAsync with a base template type and derived class instance fails silently when called from an IActorReminder

### ISSUE THREAD

**Issue by kulkani** [Microsoft/service-fabric-services-and-actors-dotnet#171]
**Describe the bug**

**To Reproduce**
Steps or code snippet to reproduce the behavior.

**Expected behavior**

**Additional context**

**Comment by amanbha** [comment]
@kulkani  please provide the repro service.

**Comment by kulkani** [comment]
Calling the below code from a IActorReminder - there is no indication of problem with serialization if we unregister from the reminder right after.

await StateManager.SetStateAsync<Exception>("exceptionName", new AggregateException(new[] { new Exception("1"), new Exception("2") }));

This causes us to think the state got saved, but when we query for it (in a different method/IActorReminder) we see an exception indicating there is no such state.

I understand that in the above line of code I am doing something incorrect w.r.t. serialization. But given that this does not throw and any other state that we save after (even if its valid) is also not persisted - its pretty confusing.

To Reproduce:
1) Call something like the below line of code from IActorReminder and unregister for the reminder right after.
await StateManager.SetStateAsync<Exception>("exceptionName", new AggregateException(new[] { new Exception("1"), new Exception("2") }));
2) In a separate method and a different invocation of the Actor call the below. This method throws saying there is no such state.
await StateManager.GetStateAsync<Exception>("exceptionName")

Expected behavior
The below should throw even if we unregistered from the reminder right away.
await StateManager.SetStateAsync<Exception>("exceptionName", new AggregateException(new[] { new Exception("1"), new Exception("2") }));

Also note that calling Actor.SaveStateAsync() forces the exception to be thrown.

**Comment by amanbha** [comment]
StateManager.SetStateAsync marks that the state is modified and needs to be saved. The state is actually saved when SaveStateAsync is called (either explicitly by user code, or by runtime after each method/reminder/timer call, when runtime calls it implicitly after the call). All states which you stored must be data contract serialzable.

**Comment by kulkani** [comment]
Yes I get that. But in this case I expected the runtime to implicitly call SaveStateAsync and throw an exception after the call - which does not happen. Does that help ?

**Comment by kulkani** [comment]
And I suspect the reason its not called implicitly is because I unregistered the reminder in that same call.

**Comment by kulkani** [comment]
I and Aman talked about this offline and looked at it more. Unregistering from IActorReminder is irrelevant. 
The serialization exception is not bubbled up to the developer in the reminder. So the developer has no way of knowing why SaveStateAsync is not working.

Either the exception needs to be bubbled up or documentation needs to say that ISyncActor.SaveStateAsync is required when setting state from a IActorReminder.

Ideally the documentation could be fixed soon and then a code change can get the exception to bubble up.

**Comment by amanbha** [comment]
The exception from implicit call to SaveStateAsync by runtime is bubbled up to the caller, which in case of Reminders is the actor runtime itself, so it catches it and traces it, since the caller is runtime itself, it cant be bubbled to user code, user can only see if it explicitly calls SaveStateAsync.
I agree the document can be more clear in this aspect, We will update the documentation to clarify this.

**Comment by kulkani** [comment]
" it cant be bubbled to user code, user can only see if it explicitly calls SaveStateAsync." - Can it not be bubbled up such that it shows up under debugging ? Or under a try catch ?

Currently there is no way for the developer to get at that exception. Having different set of rules around when SaveStateAsync is necessary versus optional (in documentation) helps but is not ideal as other people may miss that documentation and run into the same issue we did.

**Comment by amanbha** [comment]
Please ensure that the state you are saving is DataContractSerializable.
While debugging, enabling first chance exception should show this to you.

**Comment by kulkani** [comment]
"While debugging, enabling first chance exception should show this to you" - Didn't we just check on my machine that this is not the case if SaveStateAsync is not called explicitly ?

**Comment by amanbha** [comment]
Please check your VS settings.

**Comment by amanbha** [comment]
We will update the doc, keeping it open until doc is fixed.

**Comment by kulkani** [comment]
Thanks

**Comment by amanbha** [comment]
@kulkani This is the code where ActorRuntime is catching exception (and tracing it) from Reminder call and implicit SaveStateAsync.  https://github.com/Microsoft/service-fabric-services-and-actors-dotnet/blob/develop/src/Microsoft.ServiceFabric.Actors/Runtime/ActorManager.cs#L448

**Comment by kulkani** [comment]
Here are a few ways on how to bubble up the errors. 

1) Create a overload of SetStateAsync that takes a callback which is fired if there is any error in saving that piece of state. Without further thinking - this would be the least recommended option since it would be specific to each piece of state.

2) Create an overload of RegisterReminderAsync that takes a second CallBack which is called if there are any errors from the Reminder. This would be specific to a reminder.

3) There can be a virtual method OnReminderErrors on the Actor class that is invoked on Reminder errors. Devs can override this method to listen for errors across all Reminders.

We would probably prefer 3) but 2) would be ok as well.

**Comment by amanbha** [comment]
Thanks @kulkani for the suggestions, option 3 would be the best option, but it must happen under the same concurrency lock. Please update the issue title and the template with this info.
Also, please feel free to contribute to the repo to help the community.
